src/components/shared/thread/ActivityGraphFills2.js

```diff
@@ -0,0 +1,815 @@
+   * percentage into the buffer.
+   */
+  _accumulateInCategory(
+    category: IndexIntoCategoryList,
+    sampleIndex: IndexIntoSamplesTable,
+    prevSampleTime: Milliseconds,
+    sampleTime: Milliseconds,
+    nextSampleTime: Milliseconds,
+    sampleCPU: number
+  ) {
+    const {
+      rangeEnd,
+      rangeStart,
+      categoryDrawStyles,
+      xPixelsPerMs,
+      canvasPixelWidth,
+      maxThreadCPU,
+    } = this.renderedComponentSettings;
+    if (sampleTime < rangeStart || sampleTime >= rangeEnd) {
+      return;
+    }
+
+    const categoryDrawStyle = categoryDrawStyles[category];
+    const percentageBuffers = this.mutablePercentageBuffers[category];
+
+    if (categoryDrawStyle.selectedFillStyle === 'transparent') {
+      return;
+    }
+
+    const sampleStart = (prevSampleTime + sampleTime) / 2;
+    const sampleEnd = (sampleTime + nextSampleTime) / 2;
+    let pixelStart = (sampleStart - rangeStart) * xPixelsPerMs;
+    let pixelEnd = (sampleEnd - rangeStart) * xPixelsPerMs;
+    pixelStart = Math.max(0, pixelStart);
+    pixelEnd = Math.min(canvasPixelWidth - 1, pixelEnd);
+    const intPixelStart = Math.floor(pixelStart);
+    const intPixelEnd = Math.floor(pixelEnd);
+
+    // For every sample, we have a fractional interval of this sample's
+    // contribution to the graph's pixels.
+    //
+    // v       v       v       v       v       v       v       v       v
+    // +-------+-------+-----+-+-------+-------+-----+-+-------+-------+
+    // |       |       |     |///////////////////////| |       |       |
+    // |       |       |     |///////////////////////| |       |       |
+    // |       |       |     |///////////////////////| |       |       |
+    // +-------+-------+-----+///////////////////////+-+-------+-------+
+    //
+    // We have a device-pixel array of contributions. We map the fractional
+    // interval to this array of device pixels: Fully overlapping pixels are
+    // 1, and the partial overlapping pixels are the degree of overlap.
+
+    //                                 |
+    //                                 v
+    //
+    // +-------+-------+-------+-------+-------+-------+-------+-------+
+    // |       |       |       |///////////////+-------+       |       |
+    // |       |       |       |///////////////////////|       |       |
+    // |       |       +-------+///////////////////////|       |       |
+    // +-------+-------+///////////////////////////////+-------+-------+
+    const percentageBuffer = this._pickPercentageBuffer(
+      percentageBuffers,
+      sampleIndex
+    );
+    const samplePercentage = sampleCPU / maxThreadCPU;
+    for (let i = intPixelStart; i <= intPixelEnd; i++) {
+      percentageBuffer[i] += samplePercentage;
+    }
+    percentageBuffer[intPixelStart] -=
+      samplePercentage * (pixelStart - intPixelStart);
+    percentageBuffer[intPixelEnd] -=
+      samplePercentage * (1 - (pixelEnd - intPixelEnd));
```

Comment:
Nice, thanks for getting rid of the extra buffers.

As a next step, I think it would make sense to split each sample's interval up into two intervals, and use the appropriate CPU usage values for each part, as described in the document. I'll describe the idea here again, for future reference: Right now, you're using the same `samplePercentage` from `sampleStart` to `sampleEnd`. However, `samplePercentage` only describes the CPU usage between `prevSampleTime` and `sampleTime`, and not for the remaining interval after `sampleTime`. The CPU usage for the interval between `sampleTime` and `nextSampleTime` is given on the next sample. So I think you should use that value, and then make two intervals here: The first interval goes from `sampleStart` to `sampleTime` and should use the "left" CPU usage, and the second interval goes from `sampleTime` to `sampleEnd` and should use the "right" CPU usage.

Reply:
Here's an example of where splitting into two intervals would help: [Example profile](https://deploy-preview-3098--perf-html.netlify.app/public/1z57w2fvfptng9f1dgkbqmat4ef9ecw54arvagg/calltree/?globalTrackOrder=6-0-1-2-3-4-5&hiddenGlobalTracks=6-0-1-2-3-5&hiddenLocalTracksByPid=4300-1-2~1536-0~2068-2-5~988-2-0-1-3~12444-1&localTrackOrderByPid=4300-4-5-0-1-2-3-6~1536-1-0-2~8596-1-0-2~2068-6-0-1-2-3-4-5-7-8~988-2-0-1-3~12444-2-0-1-3~&range=5284m6602&thread=16&v=5)
We have a blue "Network" sample with ~90% CPU, a long gap, and then a gray "Other" sample with 0% CPU. The blue Network fill is extended halfway into the long gap, and makes it look like we were using 90% CPU during that half of the gap. But if we split the network sample into a short 90% CPU interval and a long 0% CPU interval, as I suggested in the previous comment, then the graph will correctly show 0% for the entire gap.

Reply:
Thanks for the recommendation on splitting the interval into two. I definitely want to do that but since this PR got pretty big, I'm thinking about doing it on a follow-up PR. I filed #3176 for this.